File: tests/typesafety/python/json_parsed.py
# expect: passes
# run: yes
# title: One parse at the door, and the Json type stops at it
# improves: json_blob
from dataclasses import dataclass
from typing import NewType

Pence = NewType("Pence", int)


@dataclass(frozen=True)
class Line:
    item: str
    pence: Pence
    quantity: int


def parse_line(raw: object) -> Line:
    match raw:
        case {"item": str(item), "pence": int(pence), "quantity": int(quantity)}:
            return Line(item, Pence(pence), quantity)
        case _:
            raise ValueError(f"not a line: {raw!r}")


def line_total(line: Line) -> Pence:
    return Pence(line.pence * line.quantity)


def invoice_total(lines: list[Line]) -> Pence:
    return Pence(sum(line_total(line) for line in lines))


def is_large(lines: list[Line]) -> bool:
    return invoice_total(lines) > 1000


basket = [parse_line({"item": "saddle", "pence": 155, "quantity": 4})]
assert invoice_total(basket) == 620
assert is_large(basket) is False

File: tests/typesafety/typescript/json_parsed.ts
// expect: passes

declare const penceBrand: unique symbol;
type Pence = number & { readonly [penceBrand]: true };

type Line = { readonly item: string; readonly pence: Pence; readonly quantity: number };

function parseLine(raw: unknown): Line {
  if (
    typeof raw !== "object" ||
    raw === null ||
    !("item" in raw) ||
    !("pence" in raw) ||
    !("quantity" in raw) ||
    typeof raw.item !== "string" ||
    typeof raw.pence !== "number" ||
    typeof raw.quantity !== "number"
  ) {
    throw new Error("not a line");
  }
  return { item: raw.item, pence: raw.pence as Pence, quantity: raw.quantity };
}

function lineTotal(line: Line): Pence {
  return (line.pence * line.quantity) as Pence;
}

export function invoiceTotal(lines: readonly Line[]): Pence {
  return lines.reduce((sum, line) => sum + lineTotal(line), 0) as Pence;
}

export function isLarge(lines: readonly Line[]): boolean {
  return invoiceTotal(lines) > 1000;
}

export const basket = [parseLine({ item: "saddle", pence: 155, quantity: 4 })];
